Indicators/AutoRegressiveIntegratedMovingAverage.cs

```diff
@@ -0,0 +1,300 @@
+            int period,
+            bool intercept = true
+            )
+            : base(name)
+        {
+            if (arOrder < 0 || maOrder < 0)
+            {
+                throw new ArgumentException("AR/MA orders cannot be negative.");
+            }
+
+            if (arOrder == 0)
+            {
+                throw new ArgumentException("arOrder (p) must be greater than zero for all " +
+                    "currently available fitting methods.");
+            }
+
+            if (period < Math.Max(arOrder, maOrder))
+            {
+                throw new ArgumentException("Period must exceed both arOrder and maOrder");
+            }
+
+            _arOrder = arOrder;
+            _maOrder = maOrder;
+            _diffOrder = diffOrder;
+            WarmUpPeriod = period;
+            _rollingData = new RollingWindow<double>(period);
+            _intercept = intercept;
+        }
+
+        /// <summary>
+        /// Fits an ARIMA(arOrder,diffOrder,maOrder) model of form (after differencing it <see cref="_diffOrder" /> times):
+        /// <para>
+        ///     Xₜ = c + εₜ + ΣᵢφᵢXₜ₋ᵢ +  Σᵢθᵢεₜ₋ᵢ
+        /// </para>
+        /// where the first sum has an upper limit of <see cref="_arOrder" /> and the second <see cref="_maOrder" />.
+        /// This particular constructor fits the model by means of <see cref="TwoStepFit" /> using ordinary least squares.
+        /// </summary>
+        /// <param name="arOrder">AR order -- p</param>
+        /// <param name="diffOrder">Difference order -- d</param>
+        /// <param name="maOrder">MA order -- q</param>
+        /// <param name="period">Size of the rolling series to fit onto</param>
+        public AutoRegressiveIntegratedMovingAverage(
+            int arOrder,
+            int diffOrder,
+            int maOrder,
+            int period,
+            bool intercept
+            )
+            : this($"ARIMA(({arOrder}, {diffOrder}, {maOrder}), {period}, {intercept})", arOrder, diffOrder, maOrder,
+                period)
+        {
+        }
+
+        /// <summary>
+        /// Resets this indicator to its initial state
+        /// </summary>
+        public override void Reset()
+        {
+            base.Reset();
+            _rollingData.Reset();
+        }
+
+        /// <summary>
+        /// Fits the model by means of implementing the following pseudo-code algorithm (in the form of "if{then}"):
+        /// <code>
+        /// if diffOrder > 0 {Difference data diffOrder times}
+        /// if arOrder > 0 {Fit the AR model Xₜ = ΣᵢφᵢXₜ; ε's are set to residuals from fitting this.}
+        /// if maOrder > 0 {Fit the MA parameters left over  Xₜ = c + εₜ + ΣᵢφᵢXₜ₋ᵢ +  Σᵢθᵢεₜ₋ᵢ}
+        /// Return: φ and θ estimates.
+        /// </code>
+        /// http://mbhauser.com/informal-notes/two-step-arma-estimation.pdf
+        /// </summary>
+        protected void TwoStepFit(double[] series) // Protected for any future inheritors (e.g., SARIMA)
```

Comment:
Suggest trying to separate this method into smaller logical parts it's quite complex/hard to follow. For example something like `UpdateArResidualError()`, you will know best for sure 👍